pkg/controller/service.go

```diff
@@ -82,119 +80,104 @@ func (s *ShadowServiceManager) Create(userSvc *corev1.Service) error {
+func (s *ShadowServiceManager) Delete(namespace, name string) error {
+	shadowSvcName := s.getShadowServiceName(namespace, name)
 
-	if err := s.cleanupPortMapping(oldUserSvc, newUserSvc); err != nil {
-		return nil, fmt.Errorf("unable to cleanup port mapping for service %s/%s: %w", oldUserSvc.Namespace, oldUserSvc.Name, err)
+	shadowSvc, err := s.serviceLister.Services(s.namespace).Get(shadowSvcName)
+	if kerrors.IsNotFound(err) {
+		return nil
 	}
 
-	ports, err := s.getShadowServicePorts(newUserSvc)
 	if err != nil {
-		return nil, fmt.Errorf("unable to get ports for service %s/%s: %w", newUserSvc.Namespace, newUserSvc.Name, err)
+		return err
 	}
 
-	var updatedSvc *corev1.Service
-
-	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		svc, err := s.lister.Services(s.namespace).Get(name)
-		if err != nil {
-			return fmt.Errorf("unable to get shadow service %q: %w", name, err)
-		}
+	// Removes all port mappings for the deleted service.
+	// Current shadow service ports are equal to the deleted service ports.
+	for _, svcPort := range shadowSvc.Spec.Ports {
+		s.removeServicePortMapping(namespace, name, svcPort)
+	}
 
-		newSvc := svc.DeepCopy()
-		newSvc.Spec.Ports = ports
+	return s.kubeClient.CoreV1().Services(s.namespace).Delete(shadowSvcName, &metav1.DeleteOptions{})
+}
 
-		if updatedSvc, err = s.kubeClient.CoreV1().Services(s.namespace).Update(newSvc); err != nil {
-			return fmt.Errorf("unable to update kubernetes service: %w", err)
+func (s *ShadowServiceManager) cleanupPortMappings(namespace, name string, oldShadowSvc, newShadowSvc *corev1.Service) {
+	for _, oldPort := range oldShadowSvc.Spec.Ports {
+		if !needsCleanup(newShadowSvc.Spec.Ports, oldPort) {
+			continue
 		}
 
-		return nil
-	})
-
-	if retryErr != nil {
-		return nil, fmt.Errorf("unable to update service %q: %v", name, retryErr)
+		s.removeServicePortMapping(namespace, name, oldPort)
 	}
-
-	s.log.Debugf("Updated service: %s/%s", s.namespace, name)
-
-	return updatedSvc, nil
 }
 
-// Delete deletes a shadow service associated with the given user service.
-func (s *ShadowServiceManager) Delete(userSvc *corev1.Service) error {
-	name := s.getShadowServiceName(userSvc.Name, userSvc.Namespace)
-
-	if err := s.cleanupPortMapping(userSvc, nil); err != nil {
-		return fmt.Errorf("unable to cleanup port mapping for service %s/%s: %w", userSvc.Namespace, userSvc.Name, err)
+func (s *ShadowServiceManager) removeUnusedPortMappings(shadowSvc, svc *corev1.Service) {
+	if svc == nil || shadowSvc == nil {
+		return
 	}
 
-	_, err := s.lister.Services(s.namespace).Get(name)
-	if err != nil {
-		return err
-	}
+	for _, shadowSvcPort := range shadowSvc.Spec.Ports {
+		if containsPort(svc.Spec.Ports, shadowSvcPort) {
+			continue
+		}
 
-	if err := s.kubeClient.CoreV1().Services(s.namespace).Delete(name, &metav1.DeleteOptions{}); err != nil {
-		return err
+		s.removeServicePortMapping(svc.Namespace, svc.Name, shadowSvcPort)
```

Comment:
I think there's a problem here.
If the `shadowSvcPort` is part of the svc ports, the traffic-type of the service can still be different. In this case we should call `s.removeServicePortMapping` to make sure we are not leaking ports.
If you have port 8080 reserved in the TCP state table and the service become UDP, we have to remove this port from the TCP state table.

Reply:
Ah my bad I didn't see the `cleanupPortMappings`. I find this function a bit disturbing, we are cleaning up ports according to the old traffic-type if the port has changed but not if it didn't. So we are relying on the call to `cleanupPortMapping` to cleanup what we left.

Reply:
Agreed but I haven't found a better solution yet. 

Thoughts?

Reply:
What if we tied port mappings to the `service.UID`? This value is constant and does not change on object update.

That way we could easily keep track of ports regardless of the changes made to the user service.

Reply:
I'm not sure to understand? 
The problem here is that before allocation we have to remove

Reply:
Sorry, I misunderstood.